src/wp-includes/formatting.php

```diff
@@ -914,6 +917,169 @@ function seems_utf8( $str ) {
+	 *
+	 * This is to be preferred exclusively once PHP 8.3.0 is
+	 * the minimum supported version, because even when the
+	 * status isn’t cached, it uses highly-optimized code to
+	 * validate the byte stream.
+	 */
+	return function_exists( 'mb_check_encoding' )
+		? mb_check_encoding( $bytes, 'UTF-8' )
+		: _wp_is_valid_utf8_fallback( $bytes );
+}
+
+/**
+ * Fallback mechanism for safely validating UTF-8 bytes.
+ *
+ * By implementing a raw method here the code will behave
+ * in the same way on all installed systems, regardless of
+ * what extensions are installed.
+ *
+ * > [The following table] lists all of the byte sequences that
+ * > are well-formed in UTF-8. A range of byte values such as
+ * > [A0, BF] indicates that any byte from A0 to BF (inclusive)
+ * > is well-formed in that position. Any byte value outside
+ * > of the ranges listed is ill-formed.
+ *
+ *     First Byte    Second Byte   Third Byte    Fourth Byte
+ *     [0x00,0x7F]
+ *     [0xC2,0xDF]   [0x80,0xBF]
+ *     0xE0          [0xA0,0xBF]   [0x80,0xBF]
+ *     [0xE1,0xEC]   [0x80,0xBF]   [0x80,0xBF]
+ *     0xED          [0x80,0x9F]   [0x80,0xBF]
+ *     [0xEE,0xEF]   [0x80,0xBF]   [0x80,0xBF]
+ *     0xF0          [0x90,0xBF]   [0x80,0xBF]   [0x80,0xBF]
+ *     [0xF1,0xF3]   [0x80,0xBF]   [0x80,0xBF]   [0x80,0xBF]
+ *     0xF4          [0x80,0x8F]   [0x80,0xBF]   [0x80,0xBF]
+ *
+ * Notice that all valid third and forth bytes are in the range
+ * [0x80,0xBF]. This validator takes advantage of that to only
+ * check the range of those bytes once.
+ *
+ * @see wp_is_valid_utf8
+ * @see https://lemire.me/blog/2018/05/09/how-quickly-can-you-check-that-a-string-is-valid-unicode-utf-8/
+ * @see https://www.unicode.org/versions/Unicode16.0.0/core-spec/chapter-3/#G27506
+ *
+ * @since 6.9.0
+ * @access private
+ *
+ * @param string $bytes String which might contain text encoded as UTF-8.
+ * @return bool Whether the provided bytes can decode as valid UTF-8.
+ */
+function _wp_is_valid_utf8_fallback( string $bytes ): bool {
+	$end = strlen( $bytes );
+
+	for ( $i = 0; $i < $end; $i++ ) {
+		/*
+		 * Quickly skip past US-ASCII bytes, all of which are valid UTF-8.
+		 *
+		 * This optimization step improves the speed from 10x to 100x
+		 * depending on whether the JIT has optimized the function.
+		 */
+		$i += strspn(
+			$bytes,
+			"\x00\x01\x02\x03\x04\x05\x06\x07\x08\x09\x0a\x0b\x0c\x0d\x0e\x0f" .
+			"\x10\x11\x12\x13\x14\x15\x16\x17\x18\x19\x1a\x1b\x1c\x1d\x1e\x1f" .
+			" !\"#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~\x7f",
+			$i
+		);
+		if ( $i >= $end ) {
+			break;
+		}
+
+		/*
+		 * The 0xC0 bytes are used as a fallback to truncated text because they
+		 * always indicate invalid UTF-8 just the same as a missing continuation
+		 * byte would. This saves performing length checks before decoding.
+		 */
+
+		$b1 = ord( $bytes[ $i ] );
+		$b2 = ord( $bytes[ $i + 1 ] ?? "\xC0" );
+
+		// Valid two-byte code points.
```

Comment:
A UTF-8 codepoint represented as a single byte _must be_ an ASCII byte (`U+0000` - `U+007F`). That was handled above, but it's described as an optimization. The rest of this function relies on that handling of 1-byte codepoints above because from here on 2- 3- and 4-byte codepoints are handled below.

It would be helpful to note that somewhere so it doesn't appear that one-byte handling is missing.

Reply:
clarified in e0fb4dab5d6d9115dc2172645df32b6775449a32

let me know if you find this more comprehensible. otherwise we can update it